### ISSUE TITLE
Sitemaps.xml breaks upon updates made in RankMath

### ISSUE THREAD

**Issue by ariel-integrity** [wpengine/faustjs#1885]
### Description

I'm using the suggested solution in the faust.js docs to create a sitemap for my headless site.

The only issue I've had is that when I make sitemap updates in the RankMath settings, it breaks my CPT sitemaps. 

However, if I turn the faust.js plugin off, clear cache, and turn it back on, they work. But in the meantime, I get errors on pages due to faust.js being turned off. Need a solution where I can skip that step.

### Steps to reproduce

Set up a sitemaps.xml doc in a faust.js project. 

Install RankMath and turn on the sitemap generator. Update the settings to include/not include a CPT.

See that those sitemaps, and maybe the posts/pages, break. Turn off faust.js, clear cache, and turn it back on. They work!

### Additional context

Site is up in an Atlas environment.

### @faustwp/core Version

1.2.0

### @faustwp/cli Version

1.2.1

### FaustWP Plugin Version

1.2.3

### WordPress Version

6.5.2

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

**Comment by ChrisWiegman** [comment]
Hello @ariel-integrity! 

My apologies for the late reply here. I've attempted to reproduce the issue with the free version of Rankmath and have been unable to do so.

A few things you might want to check:

1. Verify the front-end URL env variable is set correctly in `sitemap.xml.js`. 
2. Ensure your permalinks are not set to `Plain`.

Can you please verify for me that your sitemap is still broken with the above settings?

**Comment by ariel-integrity** [comment]
@ChrisWiegman Sorry for the late reply. I am facing this issue once again, but have very much narrowed it down.

My sitemaps only seem to break when I add a new post or custom post type post. Other than that, they work correctly and as expected based on the faust.js sitemaps documentation.

I have an example up and running on our dev DB - let me know if there is a secure way to send it to you.

In regards to front-end URL and permalinks:
My front-end URL is correctly set, and I am able to verify this, because the vast majority of sitemaps and links are all correct the vast majority of the time.
My permalinks are set to custom.

**Comment by josephfusco** [comment]
Hey @ariel-integrity,

Could you please clarify what you mean by “the sitemap breaks”? For example, is it producing invalid XML, are you getting a 404 error, or something else? The more details you can provide, the better we’ll be able to assist you.

----
Just a heads-up, ChrisWiegman is no longer working on this project.

**Comment by ariel-integrity** [comment]
@josephfusco Thanks for the quick update.

In regards to the specific sitemap error, I get a connection timeout. Never loads.

**Comment by josephfusco** [comment]
Hey @ariel-integrity, just wanted to provide an update that I have a proposed solution [available here](https://github.com/wpengine/faustjs/pull/1936). I was able to test it using RankMath and it seems like a viable solution. Would love to know your thoughts!

Will update this issue once a new release is cut for @faustwp/core. 🙌🏼

**Comment by josephfusco** [comment]
Hey @ariel-integrity, a new version of @faustwp/core has been released containing the new configuration option! 🎉 

## Steps to resolve a broken sitemap when using RankMath sitemaps

1. Update to [@faustwp/core@3.1.0](https://www.npmjs.com/package/@faustwp/core/v/3.1.0)
2. Update the sitemap.xml.js to use the new option, which might look something like this:
   ```js
   import { getSitemapProps } from '@faustwp/core';
   
   export default function Sitemap() {}
   
   export function getServerSideProps(ctx) {
      return getSitemapProps(ctx, {
        sitemapIndexPath: '/sitemap_index.xml', // <-- Use the RankMath sitemap URL
        frontendUrl: process.env.NEXT_PUBLIC_SITE_URL,
     });
   }
   ```
3. If the sitemap index path defined above is available, Faust.js sitemaps should start working again in your Next.js site.

**Comment by ariel-integrity** [comment]
Hi @josephfusco Thanks so much for your work on this!

I had a chance to implement it on my site at sitemap_index.xml. 

I have an issue now where the child maps are listed as:
https://www.url.com/sitemap.xml?sitemap=category-sitemap.xml

instead of 
https://www.url.com/sitemap_index.xml?sitemap=category-sitemap.xml,
which is where they now live.

Any chance you are able to recreate and help me solve?

Thanks,
Ariel

**Comment by Rotbot94** [comment]
Hi

I'm having the same issue.

I'm using custom structure on permalinks like this
`/blog/%postname%/`

Page and Post sitemaps break when updating a page/post while the Faust plugin is active in Wordpress. 
When I disable the Faust plugin and save a page/post all the sitemaps are fixed.

I'm always able to visit the sitemap index on http://wordpress.local/sitemap_index.xml and http://localhost:3000/sitemap_index.xml.

My index sitemap lists 3 sitemaps:
post-sitemap.xml
page-sitemap.xml
category-sitemap.xml

category-sitemap.xml never gets affected / never breaks when pages/posts are updated.

The steps to test this for me was the following:
- Disable Faust plugin
- Update a page
- Re-enable Faust plugin
- Confirm that Rankmath sitemaps (index and others) are available on Wordpress and Faust project
- Update page and press save
- Go to http://wordpress.local/page-sitemap.xml -  I get 404
- Go to http://localhost:3000/sitemap.xml?sitemap=page-sitemap.xml - I get 404
- Post and category sitemaps are still functional


It doesn't seem to be a frontend related error, but i'll post my sitemap.xml.ts file anyway.

```
import { getSitemapProps } from "@faustwp/core";
import { GetServerSidePropsContext } from "next/dist/types";
export default function Sitemap() {}

export function getServerSideProps(ctx: GetServerSidePropsContext) {
  return getSitemapProps(ctx, {
    sitemapIndexPath: "/sitemap_index.xml",
    frontendUrl: process.env.FRONTEND_URL,
    // Add pages only in frontend
    pages: [
      {
        path: "/login",
      },
    ],
  });
}
```
---
## Update
I did some further investigation.
I went back and disabled all my plugins to re-confirm that the Faust.js plugin was the culprit and it was still the case.

I disabled all the features on the Faust settings page and re-enabled them one-by-one and discovered that the issue relates to the feature **Enable Post and Category URL rewrites**. 

When this feature is disabled RankMath doesn't seem to have any issues generating new sitemaps.

The primary issue seems to be that the faust.js plugin attempts to do content replacement on the RankMath sitemaps which causes it to not generate the XML file.

I did some testing and it seems that rewriting URLs on RankMath is altogether unneccesary since that on the  @faustwp/core handles rewriting on its own.

For this test I have disabled **Enable Post and Category URL rewrites**.
Here you can see that  frontend handles the rewrite fine on its own.

### RankMath post sitemap - http://wordpress.local/post-sitemap.xml

<img width="1288" height="519" alt="Image" src="https://github.com/user-attachments/assets/063d44de-24fa-4ed6-9532-f9357ac450c3" />


### Frontend post sitemap - http://localhost:3000/sitemap.xml?sitemap=post-sitemap.xml

<img width="802" height="335" alt="Image" src="https://github.com/user-attachments/assets/9285275f-7625-4794-9608-4fca5ee4ae76" />
<br><br>

@ariel-integrity have you also enabled the **Enable Post and Category URL rewrites** feature?

It seems that the solution might be to prevent the Faust.js plugin from running content_replacement on RankMath sitemaps.
I'll fork the repo and return with my findings.

**Comment by Rotbot94** [comment]
>...
> It seems that the solution might be to prevent the Faust.js plugin from running content_replacement on RankMath sitemaps. I'll fork the repo and return with my findings.

This is by no means a comprehensive fix, but I found that the error occurs in /includes/replacement/callbackbacks.php on the post_link function if the function equivalent_frontend_url is run  - for some reason normalizing the url is what breaks the page/post sitemaps.

I  added an if statement to return the link as it is if the Rankmath plugin is active and if the current path is one of the sitemap paths. This solves the issue for me.

```php
function post_link( $link ) {
	global $pagenow;
	$target_pages = array(
		'admin-ajax.php',
		'index.php',
		'edit.php',
		'post.php',
		'post-new.php',
		'upload.php',
		'media-new.php',
	);

	// phpcs:ignore WordPress.Security.NonceVerification.Missing -- Nonce verified in `is_ajax_generate_permalink_request()` and `is_wp_link_ajax_request()`.
	if ( empty( $_POST ) && 'post-new.php' === $pagenow ) {
		return $link;
	}

	// Ajax requests to generate permalink.
	if ( in_array( $pagenow, $target_pages, true )
		&& is_ajax_generate_permalink_request()
	) {
		return $link;
	}

	if (
		! is_rewrites_enabled()
		|| ( function_exists( 'is_graphql_request' ) && is_graphql_request() )
		// Block editor makes REST requests on these pages to query content.
		|| ( in_array( $pagenow, $target_pages, true ) && current_user_can( 'edit_posts' ) && defined( 'REST_REQUEST' ) && REST_REQUEST )
	) {
		return $link;
	}

	// Check for wp-link-ajax requests. Used by Classic Editor when linking content.
	if ( is_wp_link_ajax_request() ) {
		return $link;
	}

    // If RankMath plugin is active and current path is a RankMath sitemap path return link as is
    if (is_plugin_active('seo-by-rank-math/rank-math.php')) {
		$request_uri = $_SERVER['REQUEST_URI'];

        if (in_array($request_uri, ['/post-sitemap.xml', '/page-sitemap.xml'], true)) {
            return $link;
        }
    }

	return equivalent_frontend_url( $link );
}
```
